R/join_keys.R

```diff
@@ -0,0 +1,668 @@
+#' @details
+#' Getter for `join_keys` that returns the relationship between pairs of datasets.
+#'
+#' @param join_keys_obj (`join_keys`) object to extract the join keys
+#' @param dataset_1 (`character`) name of first dataset.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Getter for join_keys ----
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]][["ds2"]] <- "some_col"
+#' jk[["ds1"]][["ds3"]] <- "some_col2"
+#'
+#' jk["ds1"]
+#' jk[1:2]
+#' jk[c("ds1", "ds2")]
+`[.join_keys` <- function(join_keys_obj, dataset_1 = NULL, keep_all_foreign_keys = FALSE) {
+  # Protection against missing being passed through functions
+  if (missing(dataset_1)) dataset_1 <- NULL
+
+  if (is.null(dataset_1)) {
+    return(join_keys_obj)
+  }
+
+  if (checkmate::test_integerish(dataset_1) || checkmate::test_logical(dataset_1)) {
+    dataset_1 <- names(join_keys_obj)[dataset_1]
+  }
+
+  # When retrieving a relationship pair, it will also return the symmetric key
+  new_jk <- new_join_keys()
+  queue <- dataset_1
+  bin <- character(0)
+
+  # Need to iterate on a mutating queue if subset of a dataset will also
+  #  select its parent as that parent might have relationships with others
+  #  already selected.
+  while (length(queue) > 0) {
+    ix <- queue[1]
+    queue <- queue[-1]
+
+    if (ix %in% bin) {
+      next
+    }
+    bin <- c(bin, ix)
+
+    ix_parent <- parent(join_keys_obj, ix)
+
+    if (checkmate::test_string(ix_parent, min.chars = 1) && !ix_parent %in% c(queue, bin)) {
+      queue <- c(queue, ix_parent)
+    }
+
+    ix_valid_names <- names(join_keys_obj[[ix]]) %in% c(queue, bin)
+    if (keep_all_foreign_keys) {
+      ix_valid_names <- rep(TRUE, length(names(join_keys_obj[[ix]])))
+    }
+
+    new_jk[[ix]] <- join_keys_obj[[ix]][ix_valid_names]
+
+    # Add primary key of parent
+    if (length(ix_parent) > 0) {
+      new_jk[[ix_parent]][[ix_parent]] <- join_keys_obj[[ix_parent]][[ix_parent]]
+    }
+  }
+
+  common_parents_ix <- names(parents(join_keys_obj)) %in% names(new_jk) &
+    parents(join_keys_obj) %in% names(new_jk)
+
+  if (any(common_parents_ix)) parents(new_jk) <- parents(join_keys_obj)[common_parents_ix]
+
+  new_jk
+}
+
+#' @rdname join_keys
+#' @details
+#' Setter via index directly (bypassing the need to use `join_key()`).
+#' When `dataset_2` is omitted, it will create a primary key with `dataset_2 = dataset_1`.
+#'
+#' @param value (`character` vector) value to assign.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Setter via index ----
+#'
+#' jk <- join_keys(
+#'   join_key("ds1", "ds2", "col12"),
+#'   join_key("ds3", "ds4", "col34")
+#' )
+#'
+#' # overwrites previously defined key
+#' jk["ds1"] <- list(ds2 = "(new)co12")
+#' jk["ds1"] <- list(ds3 = "col13", ds4 = "col14")
+#' jk
+#'
+#' jk[c("ds1", "ds2")] <- list(ds5 = "col*5")
+#' jk[c(1, 2)] <- list(ds5 = "col**5")
+#'
+#' # Creates primary key by only defining `dataset_1`
+#' jk["ds1"] <- "primary_key"
+#' jk
+`[<-.join_keys` <- function(join_keys_obj, dataset_1, value) {
```

Comment:
I think this method could easily throw error. In `base` it is not entirely clear what assigning to list does. You might be surprised. If you want to keep this method, to avoid code duplication please call `[[<-.join_keys` which supports the same input (from what I understand so far). 

```
a <- list(a = 1)
a["b"] <- list(c = 2)
a
# $a
# [1] 1
# $b
# [1] 2
```

Reply:
I'm a little uncomfortable making `[` the same as `[[` for a list-like object. I don't necessarily have a better idea, though.

Reply:
`[<-.join_keys` throws an error now in all cases. It suggest using `[[<-`

The getters work just as base R with `[` subsetting and outputing the same class type, `[[` returns the second level list.